var/spack/repos/builtin/packages/perl/package.py

```diff
@@ -72,6 +74,14 @@ class Perl(Package):  # Perl doesn't use Autotools, it should subclass Package
     # :5.24.1 needs zlib@:1.2.8: https://rt.cpan.org/Public/Bug/Display.html?id=120134
     depends_on('zlib@:1.2.8', when='@5.20.3:5.24.1')
 
+    # Taken from Fedora 35 perl, these fix perl for the gdbm-1.20 API change, where its
+    # https://github.com/Perl/perl5/issues/18915  return values aren't #defines anymore
+    # https://github.com/graygnuorg/perl5/commits/gdbm-20210623
+    with when('@5.34.0'):
+        patch('perl-5.35.1-Fix-GDBM_File-to-compile-with-version-1.20-and-earli.patch')
+        patch('perl-5.35.1-Raise-version-number-in-ext-GDBM_File-GDBM_File.pm.patch')
+        patch('perl-5.35.1-Fix-definition-of-ITEM_NOT_FOUND-for-pre-1.13-versio.patch')
+
```

Comment:
Can you just set an upperbound for gdbm instead of applying patches?

Reply:
> Can you just set an upperbound for gdbm instead of applying patches?

Of course I thought of this as well, but I thought it would not be desirable and that would be clear:

```bash
spack dependents --transitive perl | wc --lines
5103
```
Since nearly everything depends on perl, wouldn't that preclude everyone from actually using newer improvements in gdbm? -> https://savannah.gnu.org/news/?group=gdbm

Until today, there are these:

> #### Version 1.21 - Crash tolerance
> This version introduces an important new feature: Crash tolerance, brought to gdbm by Terence Kelly.
> #### Version 1.20 - New bucket cache
> The bucket cache support has been rewritten from scratch.
> The new code provides for **significant speed up of search operations.** ...

As spack is all about performance, and there is no perl 5.36 anywhere near, I assume having these improvements would be better than not?

Reply:
Hm yeah... perl is so wide-spread only because openssl has a build system written in perl (sigh), and big packages like cmake use openssl.

If your concern is that perl is holding back gdbm, then please create a separate PR that allows toggling the extension, since it's an optional dependency anyways, we just enable it unconditionally.

In principle spack *could* also rebuild a newer gdbm for packages that don't have the API issue, as long as down the line it was only used as a build dep (like mypkg <- gdbm and mypkg <- cmake <- curl <- openssl <- perl <- gdbm), as there can't be conflicting symbols in shared libs then, but this is not something supported in spack right now.

Also, the concretizer may find a way to avoid perl altogether if the user forces a version of gdbm that conflicts with perl (e.g. cmake <- curl <- mbedtls instead of openssl)

Reply:
Change pushed as requested.

Reply:
> Hm yeah... perl is so wide-spread only because openssl has a build system written in perl (sigh), and big packages like cmake use openssl by default.

Hmm, checking:
```py
--- a/var/spack/repos/builtin/packages/openssl/package.py
+++ b/var/spack/repos/builtin/packages/openssl/package.py
@@ -81,7 +81,7 @@ class Openssl(Package):   # Uses Fake Autotools, should subclass Package
 
     depends_on('zlib')
 
-    depends_on('perl@5.14.0:', type=('build', 'test'))
+    #depends_on('perl@5.14.0:', type=('build', 'test'))
 
     @classmethod
     def determine_version(cls, exe):

$ spack dependents --transitive perl | wc --lines
5092
```
Reducing the descendents from 5103 to 5092 is just 9 packages. Your claim does not seem to hold?

Reply:
Yeah, only was an overstatement. Also gnu packages like automake/autoconf/help2man/texinfo pull in perl, and a couple others. `spack dependents --transitive perl` shows you *possible* dependents, that means it may be possible to tweak variants and providers such that all these packages depend on perl.